File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxSum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        maxPathDown(root);
        return maxSum;
    }

    private int maxPathDown(TreeNode node) {
        if (node == null) return 0;

        int left = Math.max(maxPathDown(node.left), 0);
        int right = Math.max(maxPathDown(node.right), 0);
        maxSum = Math.max(maxSum, left + node.val + right);

        return node.val + Math.max(left, right);
    }
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    private var maxSum = Int.MIN_VALUE

    fun maxPathSum(root: TreeNode?): Int {
        maxPathDown(root)
        return maxSum
    }

    private fun maxPathDown(node: TreeNode?): Int {
        if (node == null) return 0

        val left = maxOf(maxPathDown(node.left), 0)
        val right = maxOf(maxPathDown(node.right), 0)
        maxSum = maxOf(maxSum, left + node.`val` + right)

        return node.`val` + maxOf(left, right)
    }
}